File: output/client/csharp/FSharpQt/Widgets/FileDialog.fs
module FSharpQt.Widgets.FileDialog

open System
open FSharpQt.BuilderNode
open Org.Whatever.QtTesting

type Signal =
    | CurrentChanged of path: string
    | CurrentUrlChanged of url: string
    | DirectoryEntered of dir: string
    | DirectoryUrlEntered of url: string
    | FileSelected of file: string
    | FilesSelected of files: string list
    | FilterSelected of filter: string
    | UrlSelected of url: string
    | UrlsSelected of urls: string list

type FileMode =
    | AnyFile
    | ExistingFile
    | Directory
    | ExistingFiles
with
    member internal this.QtValue =
        match this with
        | AnyFile -> FileDialog.FileMode.AnyFile
        | ExistingFile -> FileDialog.FileMode.ExistingFile
        | Directory -> FileDialog.FileMode.Directory
        | ExistingFiles -> FileDialog.FileMode.ExistingFiles

type ViewMode =
    | Detail
    | List
with
    member internal this.QtValue =
        match this with
        | Detail -> FileDialog.ViewMode.Detail
        | List -> FileDialog.ViewMode.List
    
type AcceptMode =
    | Open
    | Save
with
    member internal this.QtValue =
        match this with
        | Open -> FileDialog.AcceptMode.Open
        | Save -> FileDialog.AcceptMode.Save
    
type Attr =
    | WindowTitle of title: string
    | AcceptMode of mode: AcceptMode
    | FileMode of mode: FileMode
    | NameFilter of filter: string
    | NameFilters of filters: string list
    | MimeTypeFilters of filters: string list
    | ViewMode of mode: ViewMode
    | DefaultSuffix of suffix: string
    | Directory of dir: string

let private keyFunc = function
    | WindowTitle _ -> 0
    | AcceptMode _ -> 1
    | FileMode _ -> 2
    | NameFilter _ -> 3
    | NameFilters _ -> 4
    | MimeTypeFilters _ -> 5
    | ViewMode _ -> 6
    | DefaultSuffix _ -> 7
    | Directory _ -> 8
    
let private diffAttrs =
    genericDiffAttrs keyFunc

type private Model<'msg>(dispatch: 'msg -> unit) as this =
    let mutable dialog = FileDialog.Create(this)
    
    let mutable signalMap: Signal -> 'msg option = (fun _ -> None)
    let mutable currentMask = enum<FileDialog.SignalMask> 0
    
    let dispatcher (s: Signal) =
        match signalMap s with
        | Some msg ->
            dispatch msg
        | None ->
            ()
        
    member this.Dialog with get() = dialog
    member this.SignalMap with set value = signalMap <- value
    
    member this.SignalMask with set value =
        if value <> currentMask then
            dialog.SetSignalMask(value)
            currentMask <- value
    
    member this.ApplyAttrs(attrs: Attr list) =
        for attr in attrs do
            match attr with
            | WindowTitle title ->
                dialog.SetWindowTitle(title)
            | AcceptMode mode ->
                dialog.SetAcceptMode(mode.QtValue)
            | FileMode mode ->
                dialog.SetFileMode(mode.QtValue)
            | NameFilter filter ->
                dialog.SetNameFilter(filter)
            | NameFilters filters ->
                dialog.SetNameFilters(filters |> Array.ofList)
            | MimeTypeFilters filters ->
                dialog.SetMimeTypeFilters(filters |> Array.ofList)
            | ViewMode mode ->
                dialog.SetViewMode(mode.QtValue)
            | DefaultSuffix suffix ->
                dialog.SetDefaultSuffix(suffix)
            | Directory dir ->
                dialog.SetDirectory(dir)
                
    interface FileDialog.SignalHandler with
        member this.CurrentChanged path =
            dispatcher (CurrentChanged path)
        member this.CurrentUrlChanged url =
            dispatcher (CurrentUrlChanged url)
        member this.DirectoryEntered dir =
            dispatcher (DirectoryEntered dir)
        member this.DirectoryUrlEntered url =
            dispatcher (DirectoryUrlEntered url)
        member this.FileSelected file =
            dispatcher (FileSelected file)
        member this.FilesSelected selected =
            dispatcher (selected |> Array.toList |> FilesSelected)
        member this.FilterSelected filter =
            dispatcher (FilterSelected filter)
        member this.UrlSelected url =
            dispatcher (UrlSelected url)
        member this.UrlsSelected urls =
            dispatcher (urls |> Array.toList |> UrlsSelected)
            
    interface IDisposable with
        member this.Dispose() =
            dialog.Dispose()

let private create (attrs: Attr list) (signalMap: Signal -> 'msg option) (dispatch: 'msg -> unit) (initialMask: FileDialog.SignalMask) =
    let model = new Model<'msg>(dispatch)
    model.ApplyAttrs attrs
    model.SignalMap <- signalMap
    model.SignalMask <- initialMask
    model

let private migrate (model: Model<'msg>) (attrs: Attr list) (signalMap: Signal -> 'msg option) (signalMask: FileDialog.SignalMask) =
    model.ApplyAttrs attrs
    model.SignalMap <- signalMap
    model.SignalMask <- signalMask
    model

let private dispose (model: Model<'msg>) =
    (model :> IDisposable).Dispose()


type FileDialog<'msg>() =
    [<DefaultValue>] val mutable private model: Model<'msg>
    member val Attrs: Attr list = [] with get, set
    
    let mutable signalMask = enum<FileDialog.SignalMask> 0
    
    let mutable onCurrentChanged: (string -> 'msg) option = None
    let mutable onCurrentUrlChanged: (string -> 'msg) option = None
    let mutable onDirectoryEntered: (string -> 'msg) option = None
    let mutable onDirectoryUrlEntered: (string -> 'msg) option = None
    let mutable onFileSelected: (string -> 'msg) option = None
    let mutable onFilesSelected: (string list -> 'msg) option = None
    let mutable onFilterSelected: (string -> 'msg) option = None
    let mutable onUrlSelected: (string -> 'msg) option = None
    let mutable onUrlsSelected: (string list -> 'msg) option = None
    
    member this.OnCurrentChanged with set value =
        onCurrentChanged <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.CurrentChanged
        
    member this.OnCurrentUrlChanged with set value =
        onCurrentUrlChanged <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.CurrentUrlChanged
        
    member this.OnDirectoryEntered with set value =
        onDirectoryEntered <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.DirectoryEntered
        
    member this.OnDirectoryUrlEntered with set value =
        onDirectoryUrlEntered <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.DirectoryUrlEntered
        
    member this.OnFileSelected with set value =
        onFileSelected <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.FileSelected
        
    member this.OnFilesSelected with set value =
        onFilesSelected <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.FilesSelected
        
    member this.OnFilterSelected with set value =
        onFilterSelected <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.FilterSelected
        
    member this.OnUrlSelected with set value =
        onUrlSelected <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.UrlSelected
        
    member this.OnUrlsSelected with set value =
        onUrlsSelected <- Some value
        signalMask <- signalMask ||| FileDialog.SignalMask.UrlsSelected
            
    let signalMap (signal: Signal) =
        match signal with
        | CurrentChanged path ->
            onCurrentChanged
            |> Option.map (fun f -> f path)
        | CurrentUrlChanged url ->
            onCurrentUrlChanged
            |> Option.map (fun f -> f url)
        | DirectoryEntered dir ->
            onDirectoryEntered
            |> Option.map (fun f -> f dir)
        | DirectoryUrlEntered url ->
            onDirectoryUrlEntered
            |> Option.map (fun f -> f url)
        | FileSelected file ->
            onFileSelected
            |> Option.map (fun f -> f file)
        | FilesSelected files ->
            onFilesSelected
            |> Option.map (fun f -> f files)
        | FilterSelected filter ->
            onFilterSelected
            |> Option.map (fun f -> f filter)
        | UrlSelected url ->
            onUrlSelected
            |> Option.map (fun f -> f url)
        | UrlsSelected urls ->
            onUrlsSelected
            |> Option.map (fun f -> f urls)
            
    interface IDialogNode<'msg> with
        override this.Dependencies = []
            
        override this.Create(dispatch: 'msg -> unit) =
            this.model <- create this.Attrs signalMap dispatch signalMask
            
        override this.MigrateFrom (left: IBuilderNode<'msg>) (depsChanges: (DepsKey * DepsChange) list) =
            let left' = (left :?> FileDialog<'msg>)
            let nextAttrs =
                diffAttrs left'.Attrs this.Attrs
                |> createdOrChanged
            this.model <- migrate left'.model nextAttrs signalMap signalMask
            
        override this.Dispose() =
            (this.model :> IDisposable).Dispose()
            
        override this.Dialog =
            this.model.Dialog
            
        override this.ContentKey =
            (this :> IDialogNode<'msg>).Dialog
            
        override this.AttachedToWindow window =
            this.model.Dialog.SetParentDialogFlags(window)

File: output/server/cpp/QtTestingImpl/FileDialog.cpp
#include "generated/FileDialog.h"

#include <QFileDialog>

#include "util/SignalStuff.h"

#define THIS ((FileDialogWithHandler*)_this)

namespace FileDialog
{
    class FileDialogWithHandler : public QFileDialog {
        Q_OBJECT
    private:
        std::shared_ptr<SignalHandler> handler;
        uint32_t lastMask = 0;
        std::vector<SignalMapItem<SignalMask>> signalMap = {
            // from Dialog:
            { SignalMask::Accepted, SIGNAL(accepted()), SLOT(onAccepted()) },
            { SignalMask::Finished, SIGNAL(finished(int)), SLOT(onFinished(int)) },
            { SignalMask::Rejected, SIGNAL(rejected()), SLOT(onRejected()) },
            // FileDialog:
            { SignalMask::CurrentChanged, SIGNAL(currentChanged(QString)), SLOT(onCurrentChanged(QString)) },
            { SignalMask::CurrentUrlChanged, SIGNAL(currentUrlChanged(QUrl)), SLOT(onCurrentUrlChanged(QUrl)) },
            { SignalMask::DirectoryEntered, SIGNAL(directoryEntered(QString)), SLOT(onDirectoryEntered(QString)) },
            { SignalMask::DirectoryUrlEntered, SIGNAL(directoryUrlEntered(QUrl)), SLOT(onDirectoryUrlEntered(QUrl)) },
            { SignalMask::FileSelected, SIGNAL(fileSelected(QString)), SLOT(onFileSelected(QString)) },
            { SignalMask::FilesSelected, SIGNAL(filesSelected(QStringList)), SLOT(onFilesSelected(QStringList)) },
            { SignalMask::FilterSelected, SIGNAL(filterSelected(QString)), SLOT(onFilterSelected(QString)) },
            { SignalMask::UrlSelected, SIGNAL(urlSelected(QUrl)), SLOT(onUrlSelected(QUrl)) },
            { SignalMask::UrlsSelected, SIGNAL(urlsSelected(QList<QUrl>)), SLOT(onUrlsSelected(QList<QUrl>)) },
        };
    public:
        explicit FileDialogWithHandler(const std::shared_ptr<SignalHandler> &handler)
            : handler(handler) {}
        void setSignalMask(uint32_t newMask) {
            if (newMask != lastMask) {
                processChanges(lastMask, newMask, signalMap, this);
                lastMask = newMask;
            }
        }
    public slots:
        void onAccepted() {
            handler->accepted();
        };
        void onFinished(int result) {
            handler->finished(result);
        }
        void onRejected() {
            handler->rejected();
        }
        void onCurrentChanged(const QString &path) {
            handler->currentChanged(path.toStdString());
        }
        void onCurrentUrlChanged(const QUrl &url) {
            handler->currentUrlChanged(url.toString().toStdString());
        }
        void onDirectoryEntered(const QString &directory) {
            handler->directoryEntered(directory.toStdString());
        }
        void onDirectoryUrlEntered(const QUrl &directory) {
            handler->directoryUrlEntered(directory.toString().toStdString());
        }
        void onFileSelected(const QString &file) {
            handler->fileSelected(file.toStdString());
        }
        void onFilesSelected(const QStringList &selected) {
            std::vector<std::string> selected2;
            for (auto &str : selected) {
                selected2.push_back(str.toStdString());
            }
            handler->filesSelected(selected2);
        }
        void onFilterSelected(const QString &filter) {
            handler->filterSelected(filter.toStdString());
        }
        void onUrlSelected(const QUrl &url) {
            handler->urlSelected(url.toString().toStdString());
        }
        void onUrlsSelected(const QList<QUrl> &urls) {
            std::vector<std::string> urls2;
            for (auto &url: urls) {
                urls2.push_back(url.toString().toStdString());
            }
            handler->urlsSelected(urls2);
        }
    };

    void Handle_setAcceptMode(HandleRef _this, AcceptMode mode) {
        THIS->setAcceptMode((QFileDialog::AcceptMode)mode);
    }

    void Handle_setFileMode(HandleRef _this, FileMode mode) {
        THIS->setFileMode((QFileDialog::FileMode)mode);
    }

    void Handle_setNameFilter(HandleRef _this, std::string filter) {
        THIS->setNameFilter(QString::fromStdString(filter));
    }

    void Handle_setNameFilters(HandleRef _this, std::vector<std::string> filters) {
        QStringList qFilters;
        for (auto &filter: filters) {
            qFilters.append(QString::fromStdString(filter));
        }
        THIS->setNameFilters(qFilters);
    }

    void Handle_setMimeTypeFilters(HandleRef _this, std::vector<std::string> filters) {
        QStringList qFilters;
        for (auto &filter: filters) {
            qFilters.append(QString::fromStdString(filter));
        }
        THIS->setMimeTypeFilters(qFilters);
    }

    void Handle_setViewMode(HandleRef _this, ViewMode mode) {
        THIS->setViewMode((QFileDialog::ViewMode)mode);
    }

    void Handle_setDefaultSuffix(HandleRef _this, std::string suffix) {
        THIS->setDefaultSuffix(QString::fromStdString(suffix));
    }

    void Handle_setDirectory(HandleRef _this, std::string dir) {
        THIS->setDirectory(QString::fromStdString(dir));
    }

    void Handle_selectFile(HandleRef _this, std::string file) {
        THIS->selectFile(QString::fromStdString(file));
    }

    std::vector<std::string> Handle_selectedFiles(HandleRef _this) {
        std::vector<std::string> ret;
        for (auto &file : THIS->selectedFiles()) {
            ret.push_back(file.toStdString());
        }
        return ret;
    }

    void Handle_setSignalMask(HandleRef _this, uint32_t mask) {
        THIS->setSignalMask(mask);
    }

    void Handle_dispose(HandleRef _this) {
        delete THIS;
    }

    HandleRef create(std::shared_ptr<SignalHandler> handler) {
        return (HandleRef) new FileDialogWithHandler(handler);
    }
}

#include "FileDialog.moc"
